pkg/randchar/generator.go

```diff
@@ -1,54 +1,306 @@
+// source of randomness is set to crypto/rand.Reader.
+func WithReader(reader io.Reader) Option {
+	return func(r Rand) (Rand, error) {
+		r.reader = reader
+		return r, nil
+	}
+}
+
+// Generate returns a randomly generated slice of characters that meets the requirements of the reader.
+func (r Rand) Generate(n int) ([]byte, error) {
+	if n < r.minLen {
+		return nil, errors.New("n cannot be smaller than the minimum required length of the generator")
+	}
+
+	if n < 1 {
+		return []byte{}, nil
+	}
+
+	result := make([]byte, n)
+	i := 0
+	var err error
+	for _, min := range r.minima {
+		err = min.charset.rand(r.reader, result[i:i+min.count])
+		if err != nil {
+			return nil, err
+		}
+		i += min.count
+	}
+
+	err = r.base.rand(r.reader, result[i:])
+	if err != nil {
+		return nil, err
+	}
+
+	// Only shuffle if we've applied the minima.
+	// Otherwise, the result is already random and
+	// we can save time by skipping the shuffle step.
+	if i > 0 {
+		err = shuffle(r.reader, result)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+// shuffle randomly shuffles elements of a byte slice, using the Durstenfeld shuffle algorithm.
+func shuffle(reader io.Reader, data []byte) error {
+	var randomIndex *big.Int
+	var j int64
+	var err error
+	for i := len(data) - 1; i > 0; i-- {
+		randomIndex, err = rand.Int(reader, big.NewInt(int64(i)))
+		if err != nil {
+			return err
+		}
+		j = randomIndex.Int64()
+		data[i], data[j] = data[j], data[i]
+	}
+
+	return nil
+}
+
+// Charset is a byte slice with a set of unique characters.
+type Charset []byte
+
+// NewCharset creates a set of characters from a given byte slice, removing duplicates to ensure the random generators are not biased.
+func NewCharset(characters string) Charset {
+	uniques := make(map[byte]struct{})
+	for _, char := range []byte(characters) {
+		uniques[char] = struct{}{}
+	}
+
+	result := make([]byte, len(uniques))
+	i := 0
+	for char := range uniques {
+		result[i] = char
+		i++
+	}
+
+	return result
+}
+
+// Add merges two character sets into one, removing duplicates.
+func (cs Charset) Add(set Charset) Charset {
+	uniques := make(map[byte]struct{})
+	for _, char := range cs {
+		uniques[char] = struct{}{}
+	}
+
+	for _, char := range set {
+		uniques[char] = struct{}{}
+	}
+
+	result := make([]byte, len(uniques))
+	i := 0
+	for char := range uniques {
+		result[i] = char
+		i++
+	}
+
+	return result
+}
+
+// Subtract removes all characters from a set that match a given set of characters.
+func (cs Charset) Subtract(set Charset) Charset {
+	filter := make(map[byte]struct{})
+	for _, char := range set {
+		filter[char] = struct{}{}
+	}
+
+	result := []byte{}
+	for _, char := range cs {
+		_, exists := filter[char]
+		if !exists {
+			result = append(result, char)
+		}
+	}
+
+	return result
+}
+
+// IsSubset returns true when the character set is a subset of the given set.
+// When both sets are the same it returns true too.
```

Comment:
Yes, you're factually correct. Though IMO it's good to be explicit about these behaviors, because not everyone will have this ready top of mind to expect it to work that way.